crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1038,28 +1042,74 @@ impl<'db> TypeInferenceBuilder<'db> {
         parameter_with_default: &ast::ParameterWithDefault,
         definition: Definition<'db>,
     ) {
-        // TODO(dhruvmanila): Infer types from annotation or default expression
-        // TODO check that default is assignable to parameter type
-        self.infer_parameter_definition(&parameter_with_default.parameter, definition);
+        let ast::ParameterWithDefault {
+            parameter,
+            default,
+            range: _,
+        } = parameter_with_default;
+        let opt_default_ty = default
+            .as_ref()
+            .map(|default| definition_expression_ty(self.db, definition, default));
```

Comment:
Nit: It's uncommon to prefix `Option` values with `opt`. 

```suggestion
        let default_ty = default
            .as_deref()
            .map(|default| definition_expression_ty(self.db, definition, default));
```